protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,613 @@
+    /// @notice Keep track of users and their delegates
+    mapping(address account => address delegate) public userDelegates;
+    /// @notice Keep track of delegates and users delegated to them
+    mapping(address delegate => EnumerableSet.AddressSet) private delegateUsersSet;
+
+    constructor(
+        address _rescuer,
+        address _executor,
+        address _stakingToken,
+        address _rewardToken
+    ) TempleElevatedAccess(_rescuer, _executor){
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+    }
+
+    /**
+     * @notice Set starter of rewards distribution for the next epoch
+     * @dev If starter is address zero, anyone can call `distributeRewards` to apply and 
+     *      distribute
+      rewards for next 7 days
+     * @param _starter Starter address
+     * @dev could be:
+     * - a bot (if set to non zero address) that checks requirements are met before starting auction
+     * - or anyone if set to zero address
+     */
+    function setDistributionStarter(address _starter) external onlyElevatedAccess {
+        /// @notice Starter can be address zer0
+        distributionStarter = _starter;
+        emit DistributionStarterSet(_starter);
+    }
+
+    /**
+     * @notice Set migrator
+     * @param _migrator Migrator
+     */
+    function setMigrator(address _migrator) external override onlyElevatedAccess {
+        if (_migrator == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        migrator = _migrator;
+        emit MigratorSet(_migrator);
+    }
+
+    /**  
+     * @notice Set vote delegate for a user
+     * @param _delegate Delegate
+     */
+    function setUserVoteDelegate(address _delegate) external override {
+        if (_delegate == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        if (!delegates[_delegate]) {  revert InvalidDelegate(); }
+
+        // remove old delegate if any
+        address _userDelegate = userDelegates[msg.sender];
+        EnumerableSet.AddressSet storage prevDelegateUsersSet = delegateUsersSet[_userDelegate];
+        uint256 userBalance = _balances[msg.sender];
+
+        bool removed = prevDelegateUsersSet.remove(msg.sender);
+        // update vote weight and _delegateBalance
+        if (userBalance > 0) {
+            uint256 _prevBalance;
+            uint256 _newDelegateBalance;
+            if (removed) {
+                // update vote weight of old delegate
+                _prevBalance = _delegateBalances[_userDelegate];
+                _newDelegateBalance = _delegateBalances[_userDelegate] = _prevBalance - userBalance;
+                _updateAccountWeight(_userDelegate, _prevBalance, _newDelegateBalance, false);
+            }
+            /// @dev Reuse variables
+            _prevBalance = _delegateBalances[_delegate];
+            /// @dev `_prevBalance >= 0` because when a user sets delegate, vote wieght and `_delegateBalance` are updated for delegate
```

Comment:
sorry i put the comment at the wrong place. was for `remove` case
```
/// @dev `_prevBalance > 0` because when a user sets delegate, vote weight and `_delegateBalance` are updated for delegate
 _newDelegateBalance = _delegateBalances[_userDelegate] = _prevBalance - userBalance;
```